### PR TITLE
[Feature] Add ActionScaling transform (part 1 of #1209)

### PR DESCRIPTION
**Pull request by Drakit0**
## Summary

Adds `ActionScaling`, a new transform that addresses the first half of #1209.

`ActionScaling` is a spec-driven action normalization layer:

- Reads the bounds of the action spec and derives `loc = (high + low) / 2` and `scale = (high - low) / 2`.
- Exposes a normalized action space to the policy: `[-1, 1]` by default, or `[0, 1]` when `standard_normal=False`.
- On the inv direction (policy -> env), rescales the normalized action back to the env's bounds: `env_action = norm * scale + loc`.
- On the forward direction (env action -> normalized, used inside replay buffers / `nn.Module` chains), applies the inverse mapping.

Unlike `ObservationNorm`, the values are derived from the spec (not from rollouts) and there is no moving-average / `init_stats` path - the issue explicitly asks for a static, spec-based transform for actions.

### Unbounded handling

A `RuntimeError` is raised when the action spec is:

- A non-`ContinuousBox` spec (discrete or categorical).
- An `Unbounded` spec.
- A `Bounded` spec where any entry of `low` is `-inf` / `torch.finfo(dtype).min`, or any entry of `high` is `+inf` / `torch.finfo(dtype).max`.
- Has any entry with `high <= low`.

This covers both fully and partially unbounded specs, including the case where `Unbounded` internally uses the `finfo` extremes (which `torch.isfinite` reports as finite).

### What changed

- `torchrl/envs/transforms/_action.py`: implements `ActionScaling`, with a `_maybe_init_from_parent` hook that walks the transform chain so that `loc` / `scale` are populated lazily on the first inv call when the transform is attached to a `TransformedEnv`. Explicit `loc` / `scale` can also be passed to the constructor (useful for replay-buffer-only usage).
- `torchrl/envs/transforms/transforms.py`, `torchrl/envs/transforms/__init__.py`, `torchrl/envs/__init__.py`: re-export the new class.
- `docs/source/reference/envs_transforms.rst`: add `ActionScaling` to the autosummary list.
- `test/transforms/test_action_transforms.py`: full `TransformBase` coverage plus class-specific tests for the scaling math, per-dimension bounds, `[0, 1]` mode, unbounded / partially-unbounded / `finfo` extremes, explicit `loc`/`scale`, multi-key rejection, and a rollout-bounds invariant.

### How to test

```
pytest test/transforms/test_action_transforms.py::TestActionScaling -x
```

22 tests, all passing locally on Windows / Python 3.12 / Torch 2.8.

Closes the first half of #1209.

### CODE REVIEW

**Comment by Drakit0** [comment]
CI is blocked by "workflows awaiting approval" (first-time contributor). Could a maintainer please approve workflow runs so the full test suite can execute against this PR? Thanks!

**Comment by Drakit0** [comment]
@vmoens - would you be able to take a look when you have a moment? This PR addresses one half of #1209 (the other is #3766/#3765). I cannot use the "Request review" button as an external contributor, hence the mention. Thanks!

**Comment by Drakit0** [comment]
Pushed a lint-only fix - dropped two unused imports (one of which was leftover plumbing from an earlier draft) and ran black-22.3.0. All lint hooks (ufmt, flake8, pydocstyle, autoflake) now pass locally with the versions from `.pre-commit-config.yaml`. Tests still 22/22 green.

**Comment by Drakit0** [comment]
Heads up - the lint-only fix is on the new commit, but as a first-time contributor each new commit re-enters the "workflows awaiting approval" state. Could a maintainer please re-approve the new workflow run so the full matrix can verify? Lint, ufmt, flake8, pydocstyle and autoflake all pass locally with the versions from `.pre-commit-config.yaml`, and the 22 transform tests still pass. Thanks!

**Comment by Drakit0** [comment]
Quick triage on the Windows failure from run #25917872026 (commit `1bcb6c1`, the one before the lint fix):

```
test/objectives/test_ddpg.py::TestDDPG::test_ddpg_prioritized_weights FAILED
  assert tensor(5.1256e-06) > 1e-05
```

Final tally: `1 failed, 21166 passed, 6238 skipped`. The single failure is a numerical tolerance check on DDPG prioritized-buffer weight std (5.13e-6 vs. the 1e-5 threshold) - statistical flake driven by weight init / sampling, completely unrelated to `ActionScaling`. All 22 `TestActionScaling` tests pass on Windows, and `main`'s recent Windows runs are green, which is consistent with this being a sporadic flake on that DDPG test.

Should be cleared by a re-run; no action needed from me on this side.

File: torchrl/envs/transforms/_action.py [hunk 1]
```diff
@@ -857,3 +865,321 @@ def _transform_observation_spec(
...
+            shape=action_spec.shape,
+            device=action_spec.device,
+            dtype=action_spec.dtype,
+        )
+
+    def _apply_transform(self, action: torch.Tensor) -> torch.Tensor:
+        # env action -> normalized
+        self._check_initialized()
+        loc = self.loc.to(action.device)
+        scale = self.scale.to(action.device)
```
Comment (vmoens) [inline torchrl/envs/transforms/_action.py]: let's figure out a way to do this in a compile friendly manner. eg call `to()` only if needed (if device differs).

File: torchrl/envs/transforms/_action.py [hunk 2]
```diff
@@ -857,3 +865,321 @@ def _transform_observation_spec(
...
+        normalized = (action - loc) / scale
+        if not self.standard_normal:
+            normalized = (normalized + 1) / 2
+        return normalized
+
+    def _inv_apply_transform(self, action: torch.Tensor) -> torch.Tensor:
+        # normalized -> env action
+        self._check_initialized()
+        loc = self.loc.to(action.device)
+        scale = self.scale.to(action.device)
```
Comment (vmoens) [inline torchrl/envs/transforms/_action.py]: ditto

**Review by vmoens — COMMENTED** [review]
LGTM, just a minor thing to fix for compile friendliness

File: torchrl/envs/transforms/_action.py (hunk 1 above)
Comment (Drakit0) [reply torchrl/envs/transforms/_action.py]: Done in e9b914a0. Extracted a `_loc_scale(device)` helper that only calls `.to()` when `self.loc.device != device`. The device comparison resolves at trace time (device is static metadata, not data), so the common same-device path is copy-free and inserts no device-copy node in the compile graph. Verified with `torch.compile(fullgraph=True)` that the branch causes no graph break.

File: torchrl/envs/transforms/_action.py (hunk 2 above)
Comment (Drakit0) [reply torchrl/envs/transforms/_action.py]: Same fix applies here - both `_apply_transform` and `_inv_apply_transform` now go through `_loc_scale(device)` (e9b914a0).

File: torchrl/envs/transforms/_action.py (hunk 1 above)
Comment (Drakit0) [reply torchrl/envs/transforms/_action.py]: I just changed it!

File: torchrl/envs/transforms/_action.py (hunk 2 above)
Comment (Drakit0) [reply torchrl/envs/transforms/_action.py]: ditto

**Comment by Drakit0** [comment]
Friendly Monday bump, @vmoens - both PRs for #1209 are currently stuck behind "workflows awaiting approval", so CI hasn't been able to run on the latest commits yet:

- **#3765 (ActionScaling)**: your compile-friendliness feedback is addressed in e9b914a0 - `loc`/`scale` are now moved via a `_loc_scale(device)` helper that only calls `.to()` when the device actually differs. Verified with `torch.compile(fullgraph=True)` that the device check causes no graph break. Both review threads have replies.
- **#3766 (FlattenAction)**: no outstanding review comments; it just needs a first look.

Whenever you have a moment to approve the workflow runs (and re-review), that would be much appreciated. Thanks for the review on Friday!

**Comment by Drakit0** [comment]
CI triage - every unit-test job passed here: `tests-cpu` (3.10-3.14), all `tests-gpu` / `tests-stable-gpu` jobs, `unittests-cpu` (Windows), `tests-olddeps`, `tests-optdeps`, `unittests-gym`, `unittests-sklearn`, SOTA, tutorials and docs are all green.

The only red checks are `build-wheel-py3_10-cpu` / `upload-wheel-py3_10-cpu` on Windows - that is the Windows wheel-builder, unrelated to the diff (the Linux and macOS/aarch64 wheel builds for this PR all succeeded). A re-run should clear them.

So this PR is functionally green - the compile-friendliness fix from your review is in (e9b914a0) and verified.

**Review by vmoens — APPROVED** [review]
LGTM thanks

**Comment by Drakit0** [comment]
I tested your refactor on local and it's way cleaner than mine, tku!
